veracruz-server/src/veracruz_server_linux.rs

```diff
@@ -0,0 +1,912 @@
+            })?;
+
+            match message {
+                RuntimeManagerMessage::TLSData(buffer, alive) => {
+                    info!("{} bytes of TLS data received from Runtime Manager enclave (alive status: {}).", buffer.len(), alive);
+
+                    Ok((alive, buffer))
+                }
+                otherwise => {
+                    error!("Unexpected reply received back from Runtime Manager enclave.  Recevied: {:?}.", otherwise);
+
+                    Err(VeracruzServerError::InvalidRuntimeManagerMessage(otherwise))
+                }
+            }
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Trait implementations.
+    ////////////////////////////////////////////////////////////////////////////
+
+    /// An implementation of the `Drop` trait that forcibly kills the runtime
+    /// manager enclave, and closes the socket used for communicating with it, when
+    /// a `VeracruzServerLinux` struct is about to go out of scope.
+    impl Drop for VeracruzServerLinux {
+        fn drop(&mut self) {
+            info!("Dropping VeracruzServerLinux object, shutting down enclaves...");
+            if let Err(error) = self.teardown() {
+                error!(
+                    "Failed to forcibly kill Runtime Manager and Linux Root enclave process.  Error produced: {:?}.",
+                    error
+                );
+            }
+            info!("VeracruzServerLinux object killed.");
+        }
+    }
+
+    impl VeracruzServer for VeracruzServerLinux {
+        /// Creates a new instance of the `VeracruzServerLinux` type.  To do
+        /// this, we:
+        ///
+        /// 1. Spawn the Linux Root enclave,
+        /// 2. Establish a socket connection between us and the Linux Root enclave,
+        /// 3. Ask the Linux Root enclave to spawn a new Runtime Manager enclave,
+        /// 4. Establish a socket connection to the Runtime Manager enclave on
+        ///    the port assigned to us by the Linux Root enclave,
+        /// 4. Send initializing messages to both enclaves.
+        /// 5. Start the proxy attestation process, getting the certificate
+        ///    chain from the Linux Root Enclave (which handles this process on
+        ///    Linux).
+        /// 6. Register the resulting certificate chain with the Runtime Manager
+        ///    enclave, in preparation for TLS connections.
+        ///
+        /// Note that this process can fail for a number of reasons, e.g. the
+        /// enclaves may not be spawnable, socket connections can fail, the
+        /// initialization processes of the two enclaves may fail, and so on.
+        /// In those cases, an explicit error is returned.  Otherwise, we return
+        /// `Ok(vsl)`.
+        fn new(policy: &str) -> Result<Self, VeracruzServerError>
+        where
+            Self: Sized,
+        {
+            info!("Creating new Veracruz Server instance for Linux.");
```

Comment:
Would it make sense to abstract the "send"/"receive" code in this function to make it more digestible? It's quite long, and there is a lot of repetition in here.

Reply:
Yes, I'll do that.

Reply:
Done.